### PR TITLE
[framework] introduce repaint boundary in Opacity widgets

### LINKED ISSUE #101597: Opacity/FadeTransition should allow child widgets to avoid repainting
In the engine an opacity layer is rendered by rendering child layers, performing a render target switch, and then blending the resulting texture with the opacity. This texture created by child rendering is cached, and if the child layers have not changed (as determined by the layer identity/id), the texture can be re-used on subsequent frames even as the opacity itself updates.

Unfortunately in many circumstances the framework does not hit this fast path. For example, consider the following code which uses AnimatedSwitcher and its default FadeTransition:

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MaterialApp(
    checkerboardRasterCacheImages: true,
    home: Scaffold(
      body: Example(),
    ),
  ));
}

class Example extends StatefulWidget {
  const Example({Key? key}) : super(key: key);

  @override
  State<Example> createState() => _ExampleState();
}

class _ExampleState extends State<Example> {
  bool alternate = false;

  void _switch() {
    setState(() {
      alternate = !alternate;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        Container(height: 100),
        TextButton(onPressed: _switch, child: const Text('Switch')),
        AnimatedSwitcher(
          duration: const Duration(milliseconds: 250),
          child: Container(
            key: ValueKey<bool>(alternate),
            child: Column(
              children: <Widget>[
                Container(
                  width: 200,
                  height: 200,
                  color: Colors.red,
                  child:
                      alternate ? const Text('TITLE A') : const Text('TITLE B'),
                ),
              ],
            ),
          ),
        ),
      ],
    );
  }
}

```

If you observe this on device, you will notice that the raster cache checkerboard flashes - because each new frame gets a new entry as the children are not reused. This can also be observed in the timeline, where each frame has a RasterCachePopulate call indicating the texture created by child rendering has been recreated. Note that this is significantly easier to observe if the child rasterization is more complex.

ezgif-1-9af8f23920

If we replace the `Container` in the example above with a `RepaintBoundary`, then we get cached rendering:

ezgif-1-0b4c0c3469

To ensure that opacity performance can hit the fast path more frequently, we should update the Opacity and FadeTransition widgets to insert a RepaintBoundary child. This may have an impact on subpixel rendering. To evaluate I will run tests in google3 and report back. The expected impact of this change is improved raster times in applications that use AnimatedSwitcher, FadeTransition, and Opacity.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/101597

### CODE REVIEW

File: packages/flutter/lib/src/widgets/basic.dart [hunk 1]
```diff
@@ -278,6 +279,54 @@ class Opacity extends SingleChildRenderObjectWidget {
   /// would otherwise contribute relevant semantics.
   final bool alwaysIncludeSemantics;
 
+  /// Whether to wrap the provided child widget in a [RepaintBoundary].
+  ///
+  /// This value defaults to true. The repaint boundary allows the engine
+  /// to reuse textures on subsequent frames if the opacity is changing but
+  /// the child layer is not. If the child layers are usually changing along
+  /// with the opacity, setting this value to `false` will reduce the size of the
+  /// layer tree.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:288]: Document the default value.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 2]
```diff
@@ -278,6 +279,54 @@ class Opacity extends SingleChildRenderObjectWidget {
   /// would otherwise contribute relevant semantics.
   final bool alwaysIncludeSemantics;
 
+  /// Whether to wrap the provided child widget in a [RepaintBoundary].
+  ///
+  /// This value defaults to true. The repaint boundary allows the engine
+  /// to reuse textures on subsequent frames if the opacity is changing but
+  /// the child layer is not. If the child layers are usually changing along
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:286]: See also my note here: https://github.com/flutter/flutter/issues/101597#issuecomment-1093404620

If we decide to keep it on Opacity, we may want to link to AnimatedOpacity and FadeTransition from here and emphasize that those may be better choices if the opacity changes frequently?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/basic.dart:286]: The fact that we need a switch to turn this off makes me wonder if we should actually just omit this feature from the Opacity widget. Did you see a use case where this helped? I would have assumed that this is mostly helpful for animated opacities (and for those one should use AnimatedOpacity or FadeTransition).

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/basic.dart:286]: So this optimiztion isn't really about animated opacities, its just easier to see when the opacity is animated. Anytime you repaint anything above an opacity, whether or not the opacity or its children have changed - you will go down a much slower path if the engine thinks the child layers have been updated.

Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/basic.dart:286]: In this particular case though, I'm not really in love with this flag. I think we should just update the progress spinner to not use an opacity widget and instead paint with opacity.

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/basic.dart]: > I think we should just update the progress spinner to not use an opacity widget and instead paint with opacity.

+1 to that if that's possible.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 3]
```diff
@@ -278,6 +278,43 @@ class Opacity extends SingleChildRenderObjectWidget {
...
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget? child;
+
+  @override
+  Widget build(BuildContext context) {
+    return _Opacity(
+      opacity: opacity,
+      alwaysIncludeSemantics: alwaysIncludeSemantics,
+      child: RepaintBoundary(child: child),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:291]: uber-nit: maybe format this as:

```suggestion
      child: RepaintBoundary(
        child: child,
      ),
```

to make it clear to the casual reader (i.e. me right now) where the repaint boundary is getting introduced...

File: packages/flutter/lib/src/widgets/basic.dart [hunk 4]
```diff
@@ -278,6 +278,43 @@ class Opacity extends SingleChildRenderObjectWidget {
...
+  const _Opacity({
+    Key? key,
+    required this.opacity,
+    this.alwaysIncludeSemantics = false,
+    Widget? child,
+  }) : assert(opacity != null && opacity >= 0.0 && opacity <= 1.0),
+       assert(alwaysIncludeSemantics != null),
+       super(key: key, child: child);
+
+
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:314]: nit: remove one extra line of white space

File: packages/flutter/lib/src/widgets/implicit_animations.dart [hunk 5]
```diff
@@ -1759,7 +1759,7 @@ class _AnimatedOpacityState extends ImplicitlyAnimatedWidgetState<AnimatedOpacit
     return FadeTransition(
       opacity: _opacityAnimation,
       alwaysIncludeSemantics: widget.alwaysIncludeSemantics,
-      child: widget.child,
+      child: RepaintBoundary(child: widget.child),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/implicit_animations.dart:1762]: Same formatting nit

File: packages/flutter/lib/src/widgets/transitions.dart [hunk 6]
```diff
@@ -513,6 +513,42 @@ class FadeTransition extends SingleChildRenderObjectWidget {
...
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget? child;
+
+  @override
+  Widget build(BuildContext context) {
+    return _FadeTransition(
+      opacity: opacity,
+      alwaysIncludeSemantics: alwaysIncludeSemantics,
+      child: RepaintBoundary(child: child),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/transitions.dart:526]: and here as well

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/transitions.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/transitions.dart]: Done

File: packages/flutter/lib/src/widgets/implicit_animations.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/implicit_animations.dart]: Done

File: packages/flutter/lib/src/widgets/basic.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/basic.dart]: Done

File: packages/flutter/lib/src/widgets/basic.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/basic.dart]: Done

**Pull request merged**